tensorflow/python/ops/nn_ops.py

```diff
@@ -42,6 +43,96 @@
 local_response_normalization = gen_nn_ops.lrn
 
 
+def confusion_matrix(predictions, targets, name=None):
+  """'
+  Calculate the Confusion Matrix for a pair of prediction and target 1-D int arrays.
+
+  Considering a prediction array such as: [1, 2, 3]
+  And a target array such as: [2, 2, 3]
+
+  The confusion matrix returned would be the following one:
+  [[0, 1, 0]
+   [0, 1, 0]
+   [0, 0, 1]]
+
+  Where the matrix rows represent the prediction labels and the columns represents
+  the target labels. The confusion matrix is always a 2-D array of shape l X l, where
+  l is the number of valid labels for a given classification task.
+  Both prediction and target must be 1-D arrays of the same shape in order
+  for this function to work.
+
+  Args:
+    predictions: A 1-D array represeting the predictions for a given classification.
+    targets: A 1-D represeting the real labels for the classification task.
+
+  Returns:
+    A l X l matrix represeting the confusion matrix, where l in the number of
+    possible labels in the classification task.
+
+  Raises:
+    ValueError: If both predictions and targets are not 1-D vectors and do not
+                have the same size.
+  """
+  with ops.op_scope([predictions, targets], name,
+                    "confusion_matrix") as name:
+
+    predictions_tensor = ops.convert_to_tensor(predictions, name="inputs_tensor")
+    targets_tensor = ops.convert_to_tensor(targets, name="targets_tensor")
+
+    predictions_shape = predictions_tensor.get_shape()
+    targets_shape = targets_tensor.get_shape()
+
+    if predictions_shape.ndims != 1:
+      raise ValueError(
+        "targets is not a 1-D vector")
+
+    if targets_shape.ndims != 1:
+      raise ValueError(
+        "targets is not a 1-D vector")
+
+    if predictions_shape.num_elements != targets_shape.num_elements:
+      raise ValueError(
+        "predictions and targets must be the same size")
+
+
+    unique_labels = set()
+    index_to_label = {}
+
+    for index, value in enumerate(predictions):
+        unique_labels.add(value)
+        unique_labels.add(targets[index])
+
+    num_labels = len(unique_labels)
+
+    count = 0
+    for label in unique_labels:
+        index_to_label[label] = count
+        count += 1
+
+    values = [ 0 for i in range(num_labels * num_labels)]
+
+    for index, value in enumerate(predictions):
+        target = index_to_label[targets[index]]
+        value = index_to_label[value]
+
+        position = value * num_labels + target
+
+        values[position] += 1
+
+    indexes = [ [i, j] for i in range(num_labels) for j in range(num_labels)
+                if values[i * num_labels + j] != 0]
+    values = [ value for value in values if value != 0]
+    shape = (num_labels, num_labels)
+    dense_matrix = np.zeros(shape=(num_labels, num_labels))
+
```

Comment:
Would it make sense to add this to metrics folder instead of nn? We started
one in contrib/metrics -
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/metrics/python/
On Apr 21, 2016 3:35 PM, "lucasmoura" <notifications@github.com> wrote:

> In tensorflow/python/ops/nn_ops.py
> <https://github.com/tensorflow/tensorflow/pull/1999#discussion_r60666190>:
>
> > +    values = [ 0 for i in range(num_labels * num_labels)]
> > +
> > +    for index, value in enumerate(predictions):
> > +        target = index_to_label[targets[index]]
> > +        value = index_to_label[value]
> > +
> > +        position = value * num_labels + target
> > +
> > +        values[position] += 1
> > +
> > +    indexes = [ [i, j] for i in range(num_labels) for j in range(num_labels)
> > +                if values[i * num_labels + j] != 0]
> > +    values = [ value for value in values if value != 0]
> > +    shape = (num_labels, num_labels)
> > +    dense_matrix = np.zeros(shape=(num_labels, num_labels))
> > +
>
> I have managed to create both indices and values array with two lines of
> code. I don't know if my approach is the way you originally thought about
> in order to reduce the lines of code, but if it is not, please let me know.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/1999/files/303d2e059eefe8e545de59ccf08a21f397720a74#r60666190>
>